pkg/reconciler/mtbroker/broker.go

```diff
@@ -0,0 +1,393 @@
+			return fmt.Errorf("failed to reconcile triggers: %v", te)
+		}
+	} else {
+		// Broker is not ready, but propagate it's status to my triggers.
+		if te := r.propagateBrokerStatusToTriggers(ctx, b.Namespace, b.Name, &b.Status); te != nil {
+			return fmt.Errorf("Trigger reconcile failed: %v", te)
+		}
+	}
+	return err
+}
+
+func (r *Reconciler) reconcileKind(ctx context.Context, b *v1alpha1.Broker) pkgreconciler.Event {
+	logging.FromContext(ctx).Debug("Reconciling", zap.Any("Broker", b))
+	b.Status.InitializeConditions()
+	b.Status.ObservedGeneration = b.Generation
+
+	// 1. Trigger Channel is created for all events. Triggers will Subscribe to this Channel.
+	// 2. Filter Deployment.
+	// 4. K8s Service that points to Filter Deployment
+
+	chanMan, err := r.getChannelTemplate(ctx, b)
+	if err != nil {
+		b.Status.MarkTriggerChannelFailed("ChannelTemplateFailed", "Error on setting up the ChannelTemplate: %s", err)
+		return err
+	}
+
+	logging.FromContext(ctx).Info("Reconciling the trigger channel")
+	c, err := resources.NewChannel("trigger", b, &chanMan.template, TriggerChannelLabels(b.Name))
+	if err != nil {
+		logging.FromContext(ctx).Error(fmt.Sprintf("Failed to create Trigger Channel object: %s/%s", chanMan.ref.Namespace, chanMan.ref.Name), zap.Error(err))
+		return err
+	}
+
+	triggerChan, err := r.reconcileChannel(ctx, chanMan.inf, chanMan.ref, c, b)
+	if err != nil {
+		logging.FromContext(ctx).Error("Problem reconciling the trigger channel", zap.Error(err))
+		b.Status.MarkTriggerChannelFailed("ChannelFailure", "%v", err)
+		return fmt.Errorf("Failed to reconcile trigger channel: %v", err)
+	}
+
+	if triggerChan.Status.Address == nil {
+		logging.FromContext(ctx).Debug("Trigger Channel does not have an address", zap.Any("triggerChan", triggerChan))
+		b.Status.MarkTriggerChannelFailed("NoAddress", "Channel does not have an address.")
+		// Ok to return nil for error here, once channel address becomes available, this will get requeued.
+		return nil
+	}
+	if url := triggerChan.Status.Address.GetURL(); url.Host == "" {
+		// We check the trigger Channel's address here because it is needed to create the Ingress Deployment.
+		logging.FromContext(ctx).Debug("Trigger Channel does not have an address", zap.Any("triggerChan", triggerChan))
+		b.Status.MarkTriggerChannelFailed("NoAddress", "Channel does not have an address.")
+		// Ok to return nil for error here, once channel address becomes available, this will get requeued.
+		return nil
+	}
+	b.Status.TriggerChannel = &chanMan.ref
+	b.Status.PropagateTriggerChannelReadiness(&triggerChan.Status)
+
+	filterDeployment, err := r.deploymentLister.Deployments("knative-eventing").Get("broker-filter")
+	if err != nil {
+		logging.FromContext(ctx).Error("Problem reconciling filter Deployment", zap.Error(err))
+		b.Status.MarkFilterFailed("DeploymentFailure", "%v", err)
+		return err
+	}
+	b.Status.PropagateFilterDeploymentAvailability(filterDeployment)
```

Comment:
This assumes a particular anatomy, which isn't particularly flexible.  Given that all you really care about is whether the network endpoint is "ready", I would change this (and the same below) to examine the `Endpoints` resources for non-empty `addresses`.

This approach should also be portable to the single-tenant Broker, and is actually (very slightly) more correct because even if the Deployment is "ready" it doesn't necessarily follow that the Endpoints have observed that and been updated to reflect it.

Reply:
ack yes +1, just kept the existing behaviour for now as discussed offline :) I'll port your changes here.

Reply:
Just to close the loop: https://github.com/knative/eventing/pull/2766